File: examples/enterprise/full_production.py
"""
Full Production Setup
=====================
Complete production configuration with custom LLM, guardrails, tools,
dynamic variables, recording, dashboard, and all lifecycle callbacks.
Use this as a starting point for production deployments.

Requirements:
    pip install patter python-dotenv httpx

Environment variables (.env):
    ANTHROPIC_API_KEY    - Anthropic API key for Claude
    DEEPGRAM_API_KEY     - Deepgram key for speech-to-text
    ELEVENLABS_API_KEY   - ElevenLabs key for text-to-speech
    TWILIO_ACCOUNT_SID   - Twilio account SID
    TWILIO_AUTH_TOKEN    - Twilio auth token
    TWILIO_PHONE_NUMBER  - Your Twilio phone number (E.164 format)
    WEBHOOK_URL          - Public URL where Twilio can reach this server
    DASHBOARD_TOKEN      - Secret token for dashboard authentication
"""

import asyncio
import logging
import os
from datetime import datetime, timezone

import httpx
from dotenv import load_dotenv

load_dotenv()

from patter import Patter

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(name)s %(message)s")
logger = logging.getLogger("production")

ANTHROPIC_API_KEY = os.environ["ANTHROPIC_API_KEY"]
DASHBOARD_TOKEN = os.getenv("DASHBOARD_TOKEN", "change-me-in-production")

SYSTEM_PROMPT = (
    "You are a professional customer service agent for Acme Corp. "
    "The caller's name is {customer_name} and their account ID is {account_id}. "
    "Help with orders, billing, and account questions. "
    "Never discuss competitor products or give medical/legal advice. "
    "Keep responses concise — this is a phone call."
)


# ── Lifecycle callbacks ──────────────────────────────────────────────

async def on_call_start(call: dict) -> dict | None:
    """Log call start and optionally return variable overrides."""
    logger.info("Call started: %s from %s", call.get("call_id"), call.get("caller"))
    # Look up caller in your CRM and inject variables into the system prompt
    # customer = await crm.lookup(call["caller"])
    return {
        "customer_name": "Jane Doe",   # Replace with CRM lookup
        "account_id": "ACME-12345",    # Replace with CRM lookup
    }


async def on_call_end(call: dict) -> None:
    """Save transcript and metrics to database."""
    logger.info(
        "Call ended: %s duration=%ss turns=%s",
        call.get("call_id"),
        call.get("duration", 0),
        call.get("turns", 0),
    )
    # db.calls.insert({
    #     "call_id": call["call_id"],
    #     "caller": call.get("caller"),
    #     "duration": call["duration"],
    #     "transcript": call.get("transcript"),
    #     "cost": call.get("cost"),
    #     "ended_at": datetime.now(timezone.utc),
    # })


async def on_metrics(metrics: dict) -> None:
    """Forward metrics to your monitoring system."""
    cost = metrics.get("cost", {})
    latency = metrics.get("latency_p95", {})
    logger.info(
        "Metrics: call=%s cost=$%.4f p95_latency=%dms",
        metrics.get("call_id"),
        cost.get("total", 0),
        latency.get("total_ms", 0),
    )
    # datadog.gauge("patter.cost.total", cost["total"])
    # datadog.gauge("patter.latency.p95", latency["total_ms"])


async def on_message(data: dict) -> str:
    """Custom LLM handler with fallback on error."""
    history = data["history"]
    messages = [{"role": m["role"], "content": m["text"]} for m in history]

    try:
        async with httpx.AsyncClient(timeout=15.0) as client:
            response = await client.post(
                "https://api.anthropic.com/v1/messages",
                headers={
                    "x-api-key": ANTHROPIC_API_KEY,
                    "anthropic-version": "2023-06-01",
                    "content-type": "application/json",
                },
                json={
                    "model": "claude-sonnet-4-20250514",
                    "max_tokens": 256,
                    "system": SYSTEM_PROMPT,
                    "messages": messages,
                },
            )
            response.raise_for_status()
            return response.json()["content"][0]["text"]
    except Exception:
        logger.exception("LLM request failed, returning fallback")
        return (
            "I'm sorry, I'm having a brief technical issue. "
            "Could you repeat that, or I can transfer you to a human agent?"
        )


# ── Initialise Patter ────────────────────────────────────────────────

phone = Patter(
    mode="local",
    twilio_sid=os.getenv("TWILIO_ACCOUNT_SID"),
    twilio_token=os.getenv("TWILIO_AUTH_TOKEN"),
    phone_number=os.getenv("TWILIO_PHONE_NUMBER"),
    webhook_url=os.getenv("WEBHOOK_URL"),
    pricing={
        "stt_per_minute": 0.0059,       # Deepgram Nova-2
        "tts_per_1k_chars": 0.30,       # ElevenLabs
        "llm_input_per_1k_tokens": 0.003,
        "llm_output_per_1k_tokens": 0.015,
        "telephony_per_minute": 0.014,   # Twilio
    },
)

agent = phone.agent(
    provider="pipeline",
    system_prompt=SYSTEM_PROMPT,
    stt=Patter.deepgram(api_key=os.environ["DEEPGRAM_API_KEY"]),
    tts=Patter.elevenlabs(api_key=os.environ["ELEVENLABS_API_KEY"], voice="aria"),
    language="en",
    first_message="Hello {customer_name}, thanks for calling Acme Corp. How can I help?",
    variables={
        "customer_name": "Valued Customer",   # Default; overridden by on_call_start
        "account_id": "unknown",
    },
    guardrails=[
        Patter.guardrail(
            name="no_competitor_mentions",
            blocked_terms=["competitor", "rival product", "switch providers"],
            replacement="I'd love to focus on how Acme can help you. What do you need?",
        ),
        Patter.guardrail(
            name="no_medical_legal",
            check=lambda text: any(
                term in text.lower()
                for term in ["diagnosis", "prescription", "legal advice", "lawsuit"]
            ),
            replacement="I'm not qualified to advise on that. Let me transfer you.",
        ),
    ],
    tools=[
        Patter.tool(
            name="check_order_status",
            description="Check the status of a customer order by order ID",
            parameters={
                "type": "object",
                "properties": {
                    "order_id": {"type": "string", "description": "The order ID"},
                },
                "required": ["order_id"],
            },
            webhook_url="https://api.yourcompany.com/orders/status",
        ),
    ],
)

# ── Start the production server ──────────────────────────────────────
if __name__ == "__main__":
    logger.info("Starting production server on port 8000...")
    logger.info("Dashboard: http://localhost:8000/dashboard")
    asyncio.run(
        phone.serve(
            agent,
            port=8000,
            recording=True,
            dashboard=True,
            dashboard_token=DASHBOARD_TOKEN,
            on_call_start=on_call_start,
            on_call_end=on_call_end,
            on_metrics=on_metrics,
            on_message=on_message,
        )
    )

File: examples/enterprise/full_production.ts
/**
 * @file full_production.ts
 * @description Complete production-ready Patter setup with all features enabled.
 *
 * This example demonstrates every major SDK capability in one file:
 *   - Pipeline mode with custom LLM (Anthropic Claude)
 *   - Tools (function calling) with local handlers
 *   - Output guardrails for compliance
 *   - Dynamic variables for per-call personalization
 *   - Call recording via Twilio
 *   - Dashboard with token authentication
 *   - Custom pricing overrides for accurate cost tracking
 *   - All lifecycle callbacks: onCallStart, onCallEnd, onMetrics, onMessage
 *
 * Prerequisites:
 *   - All credentials in .env (see below)
 *   - WEBHOOK_URL pointing to this server (e.g. ngrok or load balancer)
 *
 * Usage:
 *   npx ts-node full_production.ts
 */

import { Patter, deepgram, elevenlabs } from "getpatter";
import dotenv from "dotenv";

dotenv.config();

const phone = new Patter({
  mode: "local",
  twilioSid: process.env.TWILIO_ACCOUNT_SID!,
  twilioToken: process.env.TWILIO_AUTH_TOKEN!,
  phoneNumber: process.env.TWILIO_PHONE_NUMBER!,
  webhookUrl: process.env.WEBHOOK_URL!,
});

const agent = phone.agent({
  systemPrompt: `You are a senior account manager for {company_name}.
The caller is {caller_name} (account #{account_id}).
Help them with billing, upgrades, and technical questions.
Never discuss competitor pricing. Never share internal metrics.`,
  provider: "pipeline",
  stt: deepgram({ apiKey: process.env.DEEPGRAM_API_KEY! }),
  tts: elevenlabs({ apiKey: process.env.ELEVENLABS_API_KEY!, voice: "rachel" }),
  voice: "rachel",
  firstMessage: "Hi {caller_name}, thanks for calling {company_name}. How can I help?",
  variables: {
    company_name: "Acme Corp",
    caller_name: "Valued Customer",
    account_id: "000000",
  },
  tools: [
    {
      name: "lookup_account",
      description: "Look up a customer account by phone number or email.",
      parameters: {
        type: "object",
        properties: {
          query: { type: "string", description: "Phone number or email" },
        },
        required: ["query"],
      },
      handler: async (args) => {
        // Replace with your CRM / database lookup
        console.log(`Tool: lookup_account("${args.query}")`);
        return JSON.stringify({ name: "Jane Doe", plan: "Enterprise", balance: "$0.00" });
      },
    },
    {
      name: "create_ticket",
      description: "Create a support ticket in the ticketing system.",
      parameters: {
        type: "object",
        properties: {
          subject: { type: "string" },
          priority: { type: "string", enum: ["low", "medium", "high"] },
        },
        required: ["subject"],
      },
      handler: async (args) => {
        console.log(`Tool: create_ticket("${args.subject}", ${args.priority})`);
        return JSON.stringify({ ticketId: "TK-4821", status: "created" });
      },
    },
  ],
  guardrails: [
    {
      name: "competitor-mention",
      blockedTerms: ["competitor-x", "rival-corp"],
      replacement: "I can only discuss our own products and services.",
    },
    {
      name: "pii-filter",
      check: (text: string) => /\b\d{3}-\d{2}-\d{4}\b/.test(text),
      replacement: "I'm not able to share sensitive personal information over the phone.",
    },
  ],
});

/** Call Anthropic Claude — swap with any LLM endpoint. */
async function callLLM(transcript: string, callId: string): Promise<string> {
  try {
    const response = await fetch("https://api.anthropic.com/v1/messages", {
      method: "POST",
      headers: {
        "x-api-key": process.env.ANTHROPIC_API_KEY!,
        "anthropic-version": "2023-06-01",
        "content-type": "application/json",
      },
      body: JSON.stringify({
        model: "claude-sonnet-4-20250514",
        max_tokens: 300,
        system: agent.systemPrompt,
        messages: [{ role: "user", content: transcript }],
      }),
    });

    const body = (await response.json()) as Record<string, unknown>;
    const content = body.content as Array<{ text: string }>;
    return content[0].text;
  } catch (error: unknown) {
    const message = error instanceof Error ? error.message : "Unknown LLM error";
    console.error(`LLM call failed for ${callId}: ${message}`);
    return "I'm sorry, I'm experiencing a temporary issue. Please hold.";
  }
}

async function main(): Promise<void> {
  await phone.serve({
    agent,
    port: 8000,
    recording: true,
    dashboard: true,
    dashboardToken: process.env.DASHBOARD_TOKEN!,

    pricing: {
      deepgram: { pricePerMinute: 0.0059 },
      elevenlabs: { pricePerCharacter: 0.00003 },
      twilio: { pricePerMinute: 0.014 },
    },

    onCallStart: async (data) => {
      console.log(`[START] Call ${data.callId} from ${data.caller}`);
      // Return overrides to personalize the agent per-call
      // e.g. look up caller in CRM and inject their name
    },

    onCallEnd: async (data) => {
      console.log(`[END] Call ${data.callId} — ${data.duration}s, ${data.turns} turns`);
      // Save transcript and metrics to your data store
      // await db.insert("call_logs", {
      //   callId: data.callId,
      //   transcript: data.transcript,
      //   duration: data.duration,
      //   cost: data.cost,
      //   endedAt: new Date().toISOString(),
      // });
    },

    onMetrics: async (data) => {
      // Forward per-turn metrics to your monitoring stack
      // await prometheus.gauge("patter_latency_ms", data.latencyMs);
      // await prometheus.counter("patter_turns_total").inc();
      if ((data.latencyMs as number) > 3000) {
        console.warn(`[ALERT] High latency: ${data.latencyMs}ms on call ${data.callId}`);
      }
    },

    onMessage: async (data) => {
      const transcript = data.transcript as string;
      const callId = data.callId as string;
      return callLLM(transcript, callId);
    },
  });

  console.log("Production agent running on port 8000");
  console.log("Dashboard: http://localhost:8000/dashboard");
}

main();
